experimental/Schemes/AffineSchemes.jl

```diff
@@ -12,94 +12,248 @@ export is_open_embedding, is_closed_embedding, is_canonically_isomorphic, hypers
-  OO::MPolyQuoLocalizedRing{BRT, BRET, RT, RET, MST}
-  # fields for caching
-  name::String # the name of this scheme for printing
+function hypersurface_complement(X::SpecType, f::RingElem) where {SpecType<:AbsSpec{<:Any, <:MPolyQuoLocalizedRing}}
+  parent(f) == OO(X) || return hypersurface_complement(X, OO(X)(f))
+  h = lifted_numerator(f)
+  U = MPolyPowersOfElement(h)
+  W, _ = Localization(OO(X), U)
+  return Spec(W)
+end
 
-  function Spec(OO::MPolyQuoLocalizedRing{BRT, BRET, RT, RET, MST}) where {BRT, BRET, RT, RET, MST}
-    return new{BRT, BRET, RT, RET, MST}(OO)
-  end
+function hypersurface_complement(X::SpecType, f::RingElem) where {SpecType<:AbsSpec{<:Any, <:MPolyLocalizedRing}}
+  parent(f) == OO(X) || return hypersurface_complement(X, OO(X)(f))
+  h = numerator(f)
+  U = MPolyPowersOfElement(h)
+  W, _ = Localization(OO(X), U)
+  return Spec(W)
 end
 
-### Type getters
+function hypersurface_complement(X::SpecType, f::RingElem) where {SpecType<:AbsSpec{<:Any, <:MPolyRing}}
+  parent(f) == OO(X) || return hypersurface_complement(X, OO(X)(f))
+  U = MPolyPowersOfElement(f)
+  W, _ = Localization(OO(X), U)
+  return Spec(W)
+end
 
-ring_type(::Type{Spec{BRT, BRET, RT, RET, MST}}) where {BRT, BRET, RT, RET, MST} = MPolyQuoLocalizedRing{BRT, BRET, RT, RET, MST}
-ring_type(X::Spec) = ring_type(typeof(X))
+function hypersurface_complement(X::SpecType, f::RingElem) where {SpecType<:AbsSpec{<:Any, <:MPolyQuo}}
+  parent(f) == OO(X) || return hypersurface_complement(X, OO(X)(f))
+  U = MPolyPowersOfElement(lift(f))
+  W, _ = Localization(OO(X), U)
+  return Spec(W)
+end
 
-base_ring_type(X::Spec{BRT, BRET, RT, RET, MST}) where {BRT, BRET, RT, RET, MST} = BRT
-base_ring_elem_type(X::Spec{BRT, BRET, RT, RET, MST}) where {BRT, BRET, RT, RET, MST} = BRET
-mult_set_type(X::Spec{BRT, BRET, RT, RET, MST}) where {BRT, BRET, RT, RET, MST} = MST
-poly_ring_type(X::Spec{BRT, BRET, RT, RET, MST}) where {BRT, BRET, RT, RET, MST} = RT
-poly_type(X::Spec{BRT, BRET, RT, RET, MST}) where {BRT, BRET, RT, RET, MST} = RET
+function hypersurface_complement(X::SpecType, f::Vector{<:RingElem}) where {SpecType<:AbsSpec{<:Any, <:MPolyQuoLocalizedRing}}
+  all(x->(parent(x) == OO(X)), f) || return hypersurface_complement(X, OO(X).(f))
+  h = lifted_numerator.(f)
+  U = MPolyPowersOfElement(ambient_ring(X), h)
+  W, _ = Localization(OO(X), U)
+  return Spec(W)
+end
 
-base_ring_type(::Type{Spec{BRT, BRET, RT, RET, MST}}) where {BRT, BRET, RT, RET, MST} = BRT
-base_ring_elem_type(::Type{Spec{BRT, BRET, RT, RET, MST}}) where {BRT, BRET, RT, RET, MST} = BRET
-mult_set_type(::Type{Spec{BRT, BRET, RT, RET, MST}}) where {BRT, BRET, RT, RET, MST} = MST
-poly_ring_type(::Type{Spec{BRT, BRET, RT, RET, MST}}) where {BRT, BRET, RT, RET, MST} = RT
-poly_type(::Type{Spec{BRT, BRET, RT, RET, MST}}) where {BRT, BRET, RT, RET, MST} = RET
+function hypersurface_complement(X::SpecType, f::Vector{<:RingElem}) where {SpecType<:AbsSpec{<:Any, <:MPolyLocalizedRing}}
```

Comment:
Why is f a vector here? For a hypersurface complement this should (from the point of view of mathematical theory) be a single element.

Reply:
This is more for internal use and passing on factors of f whenever we already have it in partially factorized form.